lib/ast/AST.cpp

```diff
@@ -14,6 +36,121 @@ void KOREObjectSymbol::addSort(KOREObjectSort *Sort) {
   sort = Sort;
 }
 
+bool KOREObjectSymbol::operator==(KOREObjectSymbol other) const {
+  if (name != other.name || arguments.size() != other.arguments.size()) {
+    return false;
+  }
+  for (int i = 0; i < arguments.size(); i++) {
+    if (!(*arguments[i] == *other.arguments[i])) return false;
+  }
+  return true;
+}
+
+std::string KOREObjectSymbol::layoutString() const {
+  std::string result;
+  for (auto arg : arguments) {
+    auto sort = dynamic_cast<KOREObjectCompositeSort *>(arg);
+    if (sort->getName() == "Map") {
+      result.push_back('1');
+    } else if (sort->getName() == "List") {
+      result.push_back('2');
+    } else if (sort->getName() == "Set") {
+      result.push_back('3');
+    } else if (sort->getName() == "Array") {
+      result.push_back('2');
+    } else if (sort->getName() == "Int") {
+      result.push_back('4');
+    } else if (sort->getName() == "Float") {
+      result.push_back('5');
+    } else if (sort->getName() == "StringBuffer") {
+      result.push_back('6');
+    } else if (sort->getName() == "Bool") {
+      result.push_back('7');
+    } else if (sort->getName() == "MInt") {
+      assert(false && "not implemented yet: MInt");
+    } else {
+      result.push_back('0');
+    }
+  }
+  return result;
+}
+
+uint8_t KOREObjectSymbol::length() const {
+  uint8_t length = 1;
+  for (auto arg : arguments) {
+    auto sort = dynamic_cast<KOREObjectCompositeSort *>(arg);
+    if (sort->getName() == "Map") {
+      length += 3;
+    } else if (sort->getName() == "List") {
+      length += 7;
+    } else if (sort->getName() == "Set") {
+      length += 3;
+    } else if (sort->getName() == "Array") {
+      length += 7;
+    } else if (sort->getName() == "Int") {
+      length += 2;
+    } else if (sort->getName() == "Float") {
+      length += 4;
+    } else if (sort->getName() == "StringBuffer") {
+      length += 1;
+    } else if (sort->getName() == "Bool") {
+      length += 1;
+    } else if (sort->getName() == "MInt") {
+      assert(false && "not implemented yet: MInt");
+    } else {
+      length += 1;
+    }
+  }
+  return length;
+}
+
+
+bool KOREObjectSymbol::isConcrete() const {
+  for (auto sort : arguments) {
+    if (auto var = dynamic_cast<KOREObjectSortVariable *>(sort)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool KOREObjectSymbol::isPolymorphic() const {
+  for (auto sort : arguments) {
+    if (auto var = dynamic_cast<KOREObjectCompositeSort *>(sort)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool KOREObjectSymbol::isBuiltin() const {
```

Comment:
Additionally there will be a static named instance of each type, e.g.,

`static KoreObjectSymbol *DVSym = ...`  Then it can be referred to simply as `KoreObjectSymbol::DVSym`

Reply:
Unfortunately these symbols have parameters and therefore there will not be a single unique instance of them. There is in fact a potentially unbounded number of possible instantiations of these symbols. However, we can make an unordered_set of strings if you'd like?

Reply:
Well, I still see two more options:
1. Create a class just for the head of symbols
2. Use enums instead of strings.

2. Probably works better.

Reply:
Ah, I see the downside of using enums for this now, I think just add the `unordered_set` and call it good.